src/util.js

```diff
@@ -1,6 +1,43 @@
 import * as platform from 'detect-browser';
 import * as base64 from 'Base64';
 
+const RESULT_METADATA_KEY = '_stitch_metadata';
+
+/**
+ * Utility which creates a function that extracts metadata
+ * from the server in the response to a pipeline request,
+ * and attaches it to the final result after the finalizer has been applied.
+ *
+ * @param {Function} [func] optional finalizer to transform the response data
+ */
+export const collectMetadata = (func) => {
+  let metadata = {};
+  const attachMetadata = res => {
+    if (typeof res === 'object' && !Object.prototype.hasOwnProperty.call(res, RESULT_METADATA_KEY)) {
+      Object.defineProperty(
+        res,
+        RESULT_METADATA_KEY,
+        {enumerable: false, configurable: false, writable: false, value: metadata}
+      );
+    }
+    return Promise.resolve(res);
+  };
+  const captureMetadata = (data) => {
+    if (typeof data === 'object' && Object.prototype.hasOwnProperty.call(data, RESULT_METADATA_KEY)) {
+      // metadata already attached to this result, pass along
+      metadata = data[RESULT_METADATA_KEY];
```

Comment:
[q] might it be better to not rely on the closure of `metadata` and instead have attachMetadata take the metadata as an argument and produce a `attachMetadata` function?